### PR TITLE
[Feature] XCM-Emulator

### PR DESCRIPTION
**Pull request by ruseinov**
The original tool has been built by [Shaun](https://github.com/shaunxw) of Acala and lives in [here](https://github.com/shaunxw/xcm-simulator/tree/master/xcm-emulator).
Important: once this gets accepted - make sure to tip Shaun for his good work.
tip: 16PNhcMEvGbTYdCht9vLy6gj8qehRehEpb2Ef

**What is this for?**
This tool is designed for testing XCM configuration and message passing between pre-defined runtimes, e.g. Polkadot and Statemint. With up-to-date chain specs it allows for verifying whether specific XCM messages work in live networks.

**Why does it have to live in cumulus repo?**
Mainly because we want it to target the latest code and relative cumulus paths as opposed to a certain release. Otherwise it's not possible to use it for testing cumulus runtimes that target polkadot master. This copy already diverges from the original tool. Furthermore, by running test examples in CI we'll be able to verify that the tool is still up-to-date with the latest polkadot and cumulus changes. 
Initially this was introduced to [polkadot repo](https://github.com/paritytech/polkadot/pull/7037), but that requires ugly patch overrides for transitive dependencies on both cumulus and polkadot.

Future work on this PR is [here](https://github.com/paritytech/cumulus/pull/2502).

### CODE REVIEW

**Comment by ruseinov** [comment]
bot fmt

**Review by gilescope — APPROVED** [review]

**Review by bkontur — APPROVED** [review]
I would maybe just move all `yayoi` stuff to `yayoi.rs` under `example/yayoi.rs`, because we wont probably use it anyway

**Comment by ruseinov** [comment]
> I would maybe just move all `yayoi` stuff to `yayoi.rs` under `example/yayoi.rs`, because we wont probably use it anyway

I guess the idea of this being a separate crate has to do with the fact that it's similar to importing a runtime. As part of the next iteration we could completely remove `test_runtime` and just use one of the preconfigured runtimes for this.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2671144

**Comment by joepetrowski** [comment]
Do not merge yet, I'd like to review.

**Review by muharem — COMMENTED** [review]
In the PR we make xcm-emulator's polkadot/substrate/cumulus dependencies point to `master` branch. Why cant we do the same for original crate? Or may be CI jobs do it for us? We do probably something similar for releases. It could be done even with our help in the original repo.
I am in favor for a solution which will let us maintain and develop only one xcm-emulator.
I think everyone will benefit from it.

**Comment by vstam1** [comment]
I would like to argue that we can completely deprecate the xcm-simulator from the Polkadot repo. Please correct me if I'm wrong, but I see the xcm-emulator as a superset of the xcm-simulator. We could create an example with a simple relay-para setup in the xcm-emulator that allows developers to quickly test ideas. The problem now is that the Kusama runtime is pretty restrictive in it's configuration, making it hard to test simple ideas in the emulator. Creating a testing relay, like the testing para runtime would allow us to create both an xcm playground (for testing ideas and instructions) and complex xcm tests with production level runtimes (testing configurations). 

Would like to know what you think.

**Comment by ruseinov** [comment]
> I would like to argue that we can completely deprecate the xcm-simulator from the Polkadot repo. Please correct me if I'm wrong, but I see the xcm-emulator as a superset of the xcm-simulator. We could create an example with a simple relay-para setup in the xcm-emulator that allows developers to quickly test ideas. The problem now is that the Kusama runtime is pretty restrictive in it's configuration, making it hard to test simple ideas in the emulator. Creating a testing relay, like the testing para runtime would allow us to create both an xcm playground (for testing ideas and instructions) and complex xcm tests with production level runtimes (testing configurations).
> 
> Would like to know what you think.

It's not completely correct. XCM-Simulator is designed to work with mock runtimes, therefore it's more suitable for all sorts of examples that don't involve fully-fledged live runtimes. XCM-Emulator, on the other hand, is designed specifically to test said live runtimes or something very close to those. 

I would say this: 
1. Use XCM-Simulator for simple xcm examples that don't care about the transport level and complex runtime configurations.
2. Use XCM-Emulator to test the real deal aka live runtimes. There are still some caveats that have to do with the fact that we're testing things in an environment without block production, but it allows for testing quite a bit. 

The truth is - both are valuable. And I foresee that both types of test examples are going to be very useful for the community. I'm hopeful that upcoming XCM-Emulator integration-tests for assets parachains, properly documented, are going to help other parachain teams understand the configurations and flow better.

**Comment by vstam1** [comment]
Yeah I see what you mean. The thing I currently do not like is that the tools are located in different repos and that they overlap quite a bit. I was looking at the configurations of the test_runtime in this pr and the parachain runtime in the xcm-simulator. The biggest difference is the transport layer. How I understand the combination of Xcm and transport layer, is that xcm execution should be independent from the transport layer. Quickly like to check with @KiChjang and @franciscoaguirre? 

So if we can match both runtimes, and create a test_relay, like we have the test_runtime, we could replace the simulator with this test_network in the emulator. I am not sure if it is possible, but I think we should at least check it. I am in favor of having one well documented tool for both an xcm playground and full integration tests then two separate tools. I also think that if we choose to keep both tools (simulator for simple xcm examples and emulator for testing live runtime configurations) that we should move them both to cumulus. What do you think @ruseinov?

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/test-frameworks-and-utilities-for-polkadot-and-friends/2612/3

**Comment by ruseinov** [comment]
> So if we can match both runtimes, and create a test_relay, like we have the test_runtime, we could replace the simulator with this test_network in the emulator. I am not sure if it is possible, but I think we should at least check it. I am in favor of having one well documented tool for both an xcm playground and full integration tests then two separate tools. I also think that if we choose to keep both tools (simulator for simple xcm examples and emulator for testing live runtime configurations) that we should move them both to cumulus. What do you think @ruseinov?

I think that consolidation of those two tools or potential deprecation of xcm-simulator is not in scope of this PR. These two are fundamentally different, even though they have a very similar API. Looking at this code and what it does makes me think that it might actually be easier to have two tools and document them separately. When it comes to moving both tools to cumulus - I'd rather have both in polkadot when the repos are consolidated (monorepo) or XCM could potentially be a shared module not belonging to either of those. For now having this in cumulus is a necessity dictated by the dependency issues that otherwise arise.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:267]: Do these tests care about the transport? 
```suggestion
	fn para_to_para() {
```

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:308]: Where do these names come from? Why not continue with "Parachain A/B/C/..."?

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:326]: Are these hardcoded weights things that people expect to hold? People will copy these values as acceptable things to use in their XCM programs, so we should ensure that they come from a well-known weights table, and exceeding it is a breaking change.

This should also be a 2D weight.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:350]: Why do we expect execution to fail? "good enough to check if the message is received" doesn't seem very meaningful.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:23]:
```suggestion
	pub struct KusamaRelay {
```

Comment (joepetrowski) [reply xcm/xcm-emulator/example/src/lib.rs:23]: Why is this even called "kusama"? It does not seem to be reading in any actual network state. In fact, it's setting the default XCM version to 3 when it is really 2 on Kusama. So this is clearly _not_ Kusama nor an accurate representation of it.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:394]:
```suggestion
	fn send_remark_with_event_to_child_parachain_a(msg: &str, count: u32) {
```

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:419]: Why is this function specific to `ParachainA`?
```suggestion
	fn parachain_a_verify_remark_event_and_reset_all_events(received: bool) {
		ParachainA::execute_with(|| {
```

File: xcm/xcm-emulator/Cargo.toml
Comment (joepetrowski) [inline xcm/xcm-emulator/Cargo.toml:3]:
```suggestion
description = "Test kit to emulate XCM program execution."
```

File: xcm/xcm-emulator/README.md
Comment (joepetrowski) [inline xcm/xcm-emulator/README.md:5]: Not OK for documentation.

File: xcm/xcm-emulator/test-runtime/Cargo.toml
Comment (joepetrowski) [inline xcm/xcm-emulator/test-runtime/Cargo.toml:3]:
```suggestion
description = "A simple runtime for cross-chain message tests."
```

**Review by joepetrowski — CHANGES_REQUESTED** [review]

File: xcm/xcm-emulator/example/src/lib.rs
Comment (ruseinov) [reply xcm/xcm-emulator/example/src/lib.rs:23]: It's called Kusama, because it's using the Kusama runtime definition. Note that those are just the examples that came with xcm-emulator in the original repo, unchanged. Those are meant to be used as guidance for builders in terms of structure and code, nothing more. 

XCM-Emulator is not designed to be reading the chain state yet, it simply allows for using pre-defined runtime definitions to test message passing between them. I'm sure this can be further enhanced to do more advanced testing, however this PR's aim is to move XCM-Emulator into the Cumulus repo and making sure it compiles/runs with the current versions of Cumulus and Polkadot. 

Furthermore, we are setting Kusama's XCM version to `3` in [parachains-integration-tests](https://github.com/paritytech/parachains-integration-tests/tree/master/examples/statemine), those are the tests that are used for release checks, so that is consistent with what we are doing here.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (ruseinov) [reply xcm/xcm-emulator/example/src/lib.rs:308]: It's a refactoring bug, messages definitely need to be fixed.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (ruseinov) [reply xcm/xcm-emulator/example/src/lib.rs:267]: The XCM-Emulator is using XCMP/UMP sinks to pass messages around, so it's relevant. Another nice thing about having those names is that they correspond to those in `parachains-integration-tests` repo and potentially to the future parachain tests.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (ruseinov) [reply xcm/xcm-emulator/example/src/lib.rs:419]: Only because `kusama_send_rmrk` is currently sending only to `parachain_a`.

**Comment by ruseinov** [comment]
I've opened another one [here](https://github.com/paritytech/cumulus/pull/2502), which is where I'm going to fix the review comments.

**Comment by ruseinov** [comment]
@joepetrowski I have addressed your comments in [this PR](https://github.com/paritytech/cumulus/pull/2502). However, I stuck to `dmp/xcmp` test names to be consistent with those in [parachains-integration-tests repo](https://github.com/paritytech/parachains-integration-tests). If this does not work for some reason - I'm happy to apply whatever gets suggested instead.

**Comment by NachoPal** [comment]
> @joepetrowski I have addressed your comments in [this PR](https://github.com/paritytech/cumulus/pull/2502). However, I stuck to `dmp/xcmp` test names to be consistent with those in [parachains-integration-tests repo](https://github.com/paritytech/parachains-integration-tests). If this does not work for some reason - I'm happy to apply whatever gets suggested instead.

I've been thinking about it and I think that gathering test under `dmp`, `ump` and `hrmp` definitions will break user stories flow. I think it is better to create tests based on concepts and use cases, e.g _Teleport Asset_ - that would include both `dmp` and `ump` (from  Relay to Para and its way back).

**Comment by ruseinov** [comment]
> > @joepetrowski I have addressed your comments in [this PR](https://github.com/paritytech/cumulus/pull/2502). However, I stuck to `dmp/xcmp` test names to be consistent with those in [parachains-integration-tests repo](https://github.com/paritytech/parachains-integration-tests). If this does not work for some reason - I'm happy to apply whatever gets suggested instead.
> 
> I've been thinking about it and I think that gathering test under `dmp`, `ump` and `hrmp` definitions will break user stories flow. I think it is better to create tests based on concepts and use cases, e.g _Teleport Asset_ - that would include both `dmp` and `ump` (from Relay to Para and its way back).

Fair point, perhaps we should regroup those tests then. Might also be a good idea to do that for integration tests too, this way it's less confusing and more consistent.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (gilescope) [reply xcm/xcm-emulator/example/src/lib.rs:274]: These tests should work the same way when HRMP exists. I think Joe's right on this one - it's not that particular.

File: xcm/xcm-emulator/Cargo.toml
Comment (gilescope) [inline xcm/xcm-emulator/Cargo.toml]:
```suggestion
description = "Test kit to emulate XCM program execution"
```

**Comment by tonyalaribe** [comment]
bot rebase

File: xcm/xcm-emulator/README.md
Comment (gilescope) [inline xcm/xcm-emulator/README.md:5]:
```suggestion
XCM-Emulator is a tool to emulate XCM program execution using
pre-configured runtimes, including those used to run on live
networks, such as Kusama, Polkadot, Statemine et cetera.
This allows for testing cross-chain message passing, verifying
outcomes, weights and side-effects. It is faster than spinning up
a zombienet and as all the chains are in one process debugging using Clion is easy.

## Limitations

As the channels are mocked, using xcm-emulator tests to test
channel setup would not be appropriate.

## Alternatives

If you just wish to test execution of various XCM instructions
against the XCM VM then the `xcm-simulator` (in the polkadot
repo) is the perfect tool for this.
```

File: xcm/xcm-emulator/example/src/lib.rs
Comment (gilescope) [reply xcm/xcm-emulator/example/src/lib.rs]: Switched to 2d weight. A well-known weights table sounds like a good idea but I'd argue that should be for a separate follow on PR.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (gilescope) [reply xcm/xcm-emulator/example/src/lib.rs]: This has been tightened up so it needs to be an origin fail.

File: xcm/xcm-emulator/README.md
Comment (gilescope) [reply xcm/xcm-emulator/README.md]: Extended the documentation to point out that xcm-simulator also exists and when to use one over the other.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2786888

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2787359

**Review by bkontur — APPROVED** [review]

File: xcm/xcm-emulator/test-runtime/Cargo.toml
Comment (gilescope) [reply xcm/xcm-emulator/test-runtime/Cargo.toml:2]: good idea

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/cumulus/-/jobs/2788453

**Review by muharem — APPROVED** [review]
@ruseinov @gilescope thank you for docs

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:269]: This seems just to be testing `SafeCallFilter`? What is the value of "testing" a `remark_with_event`?

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:22]:
```suggestion
// Set up relay chain.
```

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:31]:
```suggestion
// Set up ParachainA.
```

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:42]:
```suggestion
// Set up ParachainB.
```

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:53]:
```suggestion
// Set up ParachainC.
```

File: xcm/xcm-emulator/example/src/lib.rs
Comment (joepetrowski) [inline xcm/xcm-emulator/example/src/lib.rs:64]:
```suggestion
// Set up a network with all the declared parachains and a relay chain.
```

File: xcm/xcm-emulator/README.md
Comment (joepetrowski) [inline xcm/xcm-emulator/README.md:23]:
```suggestion
1. Set up a relay chain:
```

File: xcm/xcm-emulator/README.md
Comment (joepetrowski) [inline xcm/xcm-emulator/README.md:48]:
```suggestion
3. Set up a network with all the declared parachains and a relay chain:
```

**Review by joepetrowski — COMMENTED** [review]
Some small things remaining.

File: xcm/xcm-emulator/README.md
Comment (ruseinov) [reply xcm/xcm-emulator/README.md:13]: It’s using `*Sink` to facilitate message passing as opposed to `xcm-simulator` that is using mocks. Opening of channels is not required though.

Comment (bkontur) [reply xcm/xcm-emulator/README.md:13]: > Opening of channels is not required though.

what channels? e.g. hrmp channel needs to be opened, else parachain's XcmRouter->XcmpQueue sending messages wont work.

If "As the channels are mocked" mean ump/dmp/hrmp, I wouldnt say they are mocked. Just to read/write mesasges from these queues is orchestrated directly here with some thread_local queues without `relay-chain-interface / relay-chain-inprocess-interface` code, so lets say it uses testing/simplified transport layer between relaychain and parachain.

File: xcm/xcm-emulator/README.md
Comment (gilescope) [reply xcm/xcm-emulator/README.md:7]: It does not test the 'physical' transport of the messages as that is substituted, but the logical transportation of the messages is tested - I.e. the routing as well as the execution of the messages in the context of the foreign chain is tested.

File: xcm/xcm-emulator/README.md
Comment (ruseinov) [reply xcm/xcm-emulator/README.md:13]: > Could rewrite as: "As the messages do not physically go through the same messaging infrastructure there is some code that is not being tested compared to using slower E2E tests. In future it may be possible to run these XCM emulated tests as E2E tests (without changes)."

Sounds perfect. It's indeed not 100% the same, but not fully mocked either.

File: xcm/xcm-emulator/example/src/lib.rs
Comment (ruseinov) [reply xcm/xcm-emulator/example/src/lib.rs:269]: Yes, it's currently hitting the `SafeCallFilter`, but once that's removed - this could be reinstated. Since the purpose of those example tests is to demonstrate common use-cases, not necessarily focusing on the outcome - it could be ok to keep it as is.

According to the discussion [here](https://github.com/paritytech/polkadot/pull/6824#issuecomment-1464913946), as far as I understand it, there's no real reason why the `remark_with_event` call is disabled in the first place, as the actual event that's stored contains a hash and not the original unbounded `Vec`, correct me if I'm wrong.

That said, there's an alternative, which is to `force_create_asset`. It's a bit of a more complicated scenario, but it works.

Comment (joepetrowski) [reply xcm/xcm-emulator/example/src/lib.rs:269]: Yes, I understand all that, but this error returns [quite early](https://github.com/paritytech/polkadot/blob/release-v0.9.42/xcm/xcm-executor/src/lib.rs#L550) in the `Transact` instruction. So the only thing this tests is that this `ensure` statement works. As the error indicates, the execution is "incomplete".

Comment (ruseinov) [reply xcm/xcm-emulator/example/src/lib.rs:269]: Right, I'll raise a PR to change that to `force_create_assets`, going to be less confusing. Though it would be nice to see `remark_with_event` whitelisted too, but not as part of this PR.

Comment (ruseinov) [reply xcm/xcm-emulator/example/src/lib.rs:269]: Fixed [here](https://github.com/paritytech/cumulus/pull/2566)

File: xcm/xcm-emulator/README.md
Comment (ruseinov) [reply xcm/xcm-emulator/README.md:13]: Fixed [here](https://github.com/paritytech/cumulus/pull/2447)

**Comment by NachoPal** [comment]
I merged https://github.com/paritytech/cumulus/pull/2593 with the following improvements (most of them aimed to ease tests development and apply DRY principle):

- `decl_test_relay_chain` , `decl_test_parachain` and `decl_test_network` modified to accept multiple struct items avoiding the need of multiple declarations of the same macro for each item.
- Multi-networks declarations are possible now (previously `xcm-emulator` was only capable of working for one network)
    - This new capability potentially unlocks Bridged XCM communication (mocked transport layer not implemented yet)
- Redefined `decl_test_relay_chain` and `decl_test_parachain` struct argument:
    - New `genesis` field expecting `Storage`. This improves the previous way of declaring the initial storage state of the test externality.
    - New `on_init` field to execute arbitrary actions during struct item initialisation.
    - New `runtime` field  (this field gathers previous struct types definitions + new ones). Mandatory `xcm-emulator` types dependancies.
    - New `pallets_extra` argument. Place to add specific runtime pallets types to be accessible through the struct item implementation.
- Proper structure for `xcm-emulator` with `RelayChain` , `Parachain` , `Network` and `NetworkComponent` traits.
- Improvement of `TestExt` trait.
- Helper methods for relay an parachains items struct implementations:
    - `parent_location()`
    - `child_location_of()`
    - `account_id_of()`
    - `account_data_of()`
    - `sovereign_account_id_of()`
    - `para_id()`
    - `fund_accounts()`
    - `events()`
- `assert_expected_events` macro. It helps checking received events with pattern matching and conditional assertion of theirs inner fields. [Example](https://github.com/paritytech/cumulus/blob/342babdf31733202d8c57abc91cd19b69c008e5f/parachains/integration-tests/emulated/assets/statemint/src/tests/reserve_transfer.rs#L34-L38).
    - Added events logger to inspect received events.
- Handy helper functions `within_threshold` & `weight_within_threshold` for events assessment.

In the PR I also added the base for the System Parachains *emulated* test suite:

- `integration-tests-common` package
    - `[lib.rs](http://lib.rs)`
        - Relay Chains, System Parachains and Networks declarations (for Polkadot & Kusama networks)
        - Parameters declaration of expected senders and receivers accounts for each chain. These are meant to be across all test to improve readability and prevent developers from declaring each time their own accounts.
    - `[constants.rs](http://constants.rs)`
        - System Parachain’s genesis definitions plus other common and Parachain’s specific constants.
- A few Statemint tests have been added as a reference of how to write tests to keep tests structure consistent across the whole System Parachains test suite.

`xcm-emulator` example has been removed. Now that `xcm-emulator` lives in Cumulus, System Parachains tests work as great examples.

Previous e2e test have been moved under `parachains/integration-tests/e2e`

_**Note**_: I will add proper documentation about how to use `xcm-emulator` once the PR has been approved and I make sure no interface changes are needed.

**Comment by gilescope** [comment]
On the whole it looks good - I think it's only as we develop tests against it that we will see where we need to tweak it. Can't wait for it to be merged.

File: parachains/integration-tests/emulated/common/src/lib.rs
Comment (NachoPal) [reply parachains/integration-tests/emulated/common/src/lib.rs:76]: > New `runtime` field (this field gathers previous struct types definitions + new ones). Mandatory xcm-emulator types dependancies.
New `pallets_extra` argument. Place to add specific runtime pallets types to be accessible through the struct item implementation.

`runtime` is a fixed and expected number of types that `xcm-emulator` lib needs as a dependancy in its implementation.
`pallets_extra` is optional an you can add as many types you want. If you do not add them `xcm-emulator` implementation will still work.

Comment (gilescope) [reply parachains/integration-tests/emulated/common/src/lib.rs:76]: So the types listed in `runtime` are all mandatory for the emulator to work and are the minimum pallets that we can get away with, and then everything else is put in `pallets_extra`?

Comment (NachoPal) [reply parachains/integration-tests/emulated/common/src/lib.rs:76]: Besides the conceptual difference. The need of having both separately is coming also from declarative macros arguments handling limitations.

**Review by joepetrowski — APPROVED** [review]

**Comment by joepetrowski** [comment]
bot merge

**Comment by NachoPal** [comment]
@bkontur @ggwpez @mordamax Could any of you please drop a medium tip here?

**Comment by mordamax** [comment]
@NachoPal sure,
@ruseinov is that the address you use? Polkadot address: 127o5MyqLCjZzggX5kTUokpvspKU4SgC1eRioJMYckVF4Ndm
We can try kusama thorough opengov if you want

**Comment by bkchr** [comment]
I think @shaunxw already got some tip for it back in the days? If not, he can just request tip on chain.

**Comment by xlc** [comment]
He got it for simulator, not emulator

**Pull request merged**
